pkg/planner/indexadvisor/indexadvisor.go

```diff
@@ -0,0 +1,247 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package indexadvisor
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/pingcap/tidb/pkg/sessionctx"
+	"github.com/pingcap/tidb/pkg/util/intest"
+	s "github.com/pingcap/tidb/pkg/util/set"
+	"go.uber.org/zap"
+)
+
+// TestKey is the key for test context.
+func TestKey(key string) string {
+	return "__test_index_advisor_" + key
+}
+
+// Option is the option for the index advisor.
+type Option struct {
+	MaxNumIndexes int
+	MaxIndexWidth int
+	SpecifiedSQLs []string
+}
+
+// AdviseIndexes is the only entry point for the index advisor.
+func AdviseIndexes(ctx context.Context, sctx sessionctx.Context,
+	option *Option) (results []*Recommendation, err error) {
+	if ctx == nil || sctx == nil || option == nil {
+		return nil, errors.New("nil input")
+	}
+
+	l().Info("start to recommend indexes")
+	defer func() {
+		if r := recover(); r != nil {
+			l().Error("panic in AdviseIndexes", zap.Any("recover", r))
+			err = fmt.Errorf("panic in AdviseIndexes: %v", r)
+		}
+	}()
+
+	// prepare what-if optimizer
+	opt := NewOptimizer(sctx)
+	l().Info("what-if optimizer prepared")
```

Comment:
```l``` is too hard to understand this meaning. We can call it ```indexAdvisorLog```

Reply:
Thanks for this suggestion, I've updated it.